src/test/regress/citus_tests/common.py

```diff
@@ -315,8 +347,711 @@ def eprint(*args, **kwargs):
+            autocommit=autocommit,
+            **kwargs,
+        ) as conn:
+            with conn.cursor() as cur:
+                yield cur
+
+    @asynccontextmanager
+    async def acur(self, **kwargs):
+        """Open an asynchronous psycopg cursor to this server
+
+        The connection and the cursors automatically close once you leave the
+        "async with" block
+        """
+        async with await self.aconn(**kwargs) as conn:
+            async with conn.cursor() as cur:
+                yield cur
+
+    def sql(self, query, params=None, **kwargs):
+        """Run an SQL query
+
+        This opens a new connection and closes it once the query is done
+        """
+        with self.cur(**kwargs) as cur:
+            cur.execute(query, params=params)
+
+    def sql_value(self, query, params=None, allow_empty_result=False, **kwargs):
+        """Run an SQL query that returns a single cell and return this value
+
+        This opens a new connection and closes it once the query is done
+        """
+        with self.cur(**kwargs) as cur:
+            cur.execute(query, params=params)
+            result = cur.fetchall()
+
+            if allow_empty_result and len(result) == 0:
+                return None
+
+            assert len(result) == 1
+            assert len(result[0]) == 1
+            value = result[0][0]
+            return value
+
+    def asql(self, query, **kwargs):
+        """Run an SQL query in asynchronous task
+
+        This opens a new connection and closes it once the query is done
+        """
+        return asyncio.ensure_future(self.asql_coroutine(query, **kwargs))
+
+    async def asql_coroutine(
+        self, query, params=None, **kwargs
+    ) -> typing.Optional[typing.List[typing.Any]]:
+        async with self.acur(**kwargs) as cur:
+            await cur.execute(query, params=params)
+            try:
+                return await cur.fetchall()
+            except psycopg.ProgrammingError as e:
+                if "the last operation didn't produce a result" == str(e):
+                    return None
+                raise
+
+    def psql(self, query, **kwargs):
+        """Run an SQL query using psql instead of psycopg
+
+        This opens a new connection and closes it once the query is done
+        """
+
+        self.set_default_connection_options(kwargs)
+        connect_options = " ".join([f"{k}={v}" for k, v in kwargs.items()])
+
+        run(
+            ["psql", f"port={self.port} {connect_options}", "-c", query],
+            shell=False,
+            silent=True,
+        )
+
+    def poll_query_until(self, query, params=None, **kwargs):
+        start = datetime.now()
+
+        while start + TIMEOUT_DEFAULT > datetime.now():
+            if self.sql_value(query, params=params, allow_empty_result=True, **kwargs):
+                return
+
+            time.sleep(0.1)
+
+        raise Exception("Timeout reached while polling query")
+
+    def lsn(self, mode):
```

Comment:
So this QueryRunner class is:
1. An artifact from the PgBouncer test suite. There both a pgbouncer instance and a postgres instance can handle queries. So being able to call the same querying related methods is very useful. That's why I created QueryRunner as a mixin class.
2. A hopeful idea for the future where you the `Cluster` class also extends the QueryRunner mixin. 

Maybe I should just implement 2.

Reply:
Writing this last comment made me realize that even some query methods don't make sense at the cluster level, specifically the `lsn` method you added this comment to. Because it will return different values per node anyway.